src/cbmc/bmc_util.cpp

```diff
@@ -0,0 +1,267 @@
+#include <goto-symex/memory_model_pso.h>
+#include <goto-symex/slice.h>
+#include <goto-symex/slice_by_trace.h>
+#include <goto-symex/symex_target_equation.h>
+
+#include <linking/static_lifetime_init.h>
+
+#include <solvers/prop/prop_conv.h>
+
+#include <util/make_unique.h>
+#include <util/ui_message.h>
+
+void build_error_trace(
+  goto_tracet &goto_trace,
+  const namespacet &ns,
+  const symex_target_equationt &symex_target_equation,
+  const prop_convt &prop_conv,
+  ui_message_handlert &ui_message_handler)
+{
+  messaget msg(ui_message_handler);
+  msg.status() << "Building error trace" << messaget::eom;
+
+  build_goto_trace(symex_target_equation, prop_conv, ns, goto_trace);
+}
+
+void output_error_trace(
+  const goto_tracet &goto_trace,
+  const namespacet &ns,
+  const trace_optionst &trace_options,
+  ui_message_handlert &ui_message_handler)
+{
+  messaget msg(ui_message_handler);
+  switch(ui_message_handler.get_ui())
+  {
+  case ui_message_handlert::uit::PLAIN:
+    msg.result() << "Counterexample:" << messaget::eom;
+    show_goto_trace(msg.result(), ns, goto_trace, trace_options);
+    msg.result() << messaget::eom;
+    break;
+
+  case ui_message_handlert::uit::XML_UI:
+  {
+    xmlt xml;
+    convert(ns, goto_trace, xml);
+    msg.status() << xml;
+  }
+  break;
+
+  case ui_message_handlert::uit::JSON_UI:
+  {
+    json_stream_objectt &json_result =
+      ui_message_handler.get_json_stream().push_back_stream_object();
+    const goto_trace_stept &step = goto_trace.steps.back();
+    json_result["property"] =
+      json_stringt(step.pc->source_location.get_property_id());
+    json_result["description"] =
+      json_stringt(step.pc->source_location.get_comment());
+    json_result["status"] = json_stringt("failed");
+    json_stream_arrayt &json_trace =
+      json_result.push_back_stream_array("trace");
+    convert<json_stream_arrayt>(ns, goto_trace, json_trace, trace_options);
+  }
+  break;
+  }
+}
+
+/// outputs witnesses in graphml format
+void output_graphml(
+  safety_checkert::resultt result,
+  const goto_tracet &goto_trace,
+  const symex_target_equationt &symex_target_equation,
+  const namespacet &ns,
+  const optionst &options)
+{
+  const std::string graphml = options.get_option("graphml-witness");
+  if(graphml.empty())
+    return;
+
+  graphml_witnesst graphml_witness(ns);
+  if(result == safety_checkert::resultt::UNSAFE)
+    graphml_witness(goto_trace);
+  else if(result == safety_checkert::resultt::SAFE)
+    graphml_witness(symex_target_equation);
+  else
+    return;
+
+  if(graphml == "-")
+    write_graphml(graphml_witness.graph(), std::cout);
+  else
+  {
+    std::ofstream out(graphml);
+    write_graphml(graphml_witness.graph(), out);
+  }
+}
+
+void convert_symex_target_equation(
+  symex_target_equationt &equation,
+  prop_convt &prop_conv,
+  message_handlert &message_handler)
+{
+  messaget msg(message_handler);
+  msg.status() << "converting SSA" << messaget::eom;
```

Comment:
Might be a good time to improve this rubbish message :) "Converting VCCs to propositional formula"?

Reply:
Although, this isn't a very useful utility-- print a message then call another public function? Probably just inline this?